pinot-transport/src/main/java/com/linkedin/pinot/requestHandler/BrokerRequestHandler.java

```diff
@@ -125,107 +120,108 @@ public BrokerRequestHandler(RoutingTable table, TimeBoundaryService timeBoundary
-      return brokerResponse;
+      return BrokerResponseFactory.getBrokerResponseWithException(DEFAULT_BROKER_RESPONSE_TYPE,
+          QueryException.getException(QueryException.PQL_PARSING_ERROR, e));
     }
-
-    _brokerMetrics.addMeteredQueryValue(brokerRequest, BrokerMeter.QUERIES, 1);
-
-    final long requestCompilationTime = System.nanoTime() - startTime;
-    _brokerMetrics.addPhaseTiming(brokerRequest, BrokerQueryPhase.REQUEST_COMPILATION, requestCompilationTime);
-    final ScatterGatherStats scatterGatherStats = new ScatterGatherStats();
-
+    String tableName = brokerRequest.getQuerySource().getTableName();
     try {
       validateRequest(brokerRequest);
     } catch (Exception e) {
-      BrokerResponse brokerResponse = new BrokerResponseNative();
-      brokerResponse.setExceptions(Arrays.asList(QueryException.getException(QueryException.QUERY_VALIDATION_ERROR, e)));
-
-      LOGGER.error("Validation error on query: {}", pql);
-      return brokerResponse;
-    }
-
-    final BrokerResponse resp =
-        _brokerMetrics.timeQueryPhase(brokerRequest, BrokerQueryPhase.QUERY_EXECUTION, new Callable<BrokerResponse>() {
-          @Override
-          public BrokerResponse call() throws Exception {
-            final BucketingSelection bucketingSelection = getBucketingSelection(brokerRequest);
-            return (BrokerResponse) processBrokerRequest(brokerRequest, bucketingSelection, scatterGatherStats,
-                requestId);
-          }
-        });
-
-    // Query processing time is the total time spent in all steps include query parsing, scatter/gather, ser/de etc.
-    long queryProcessingTimeInNanos = System.nanoTime() - startTime;
-    long queryProcessingTimeInMillis = TimeUnit.MILLISECONDS.convert(queryProcessingTimeInNanos, TimeUnit.NANOSECONDS);
-    resp.setTimeUsedMs(queryProcessingTimeInMillis);
-
-    LOGGER.debug("Broker Response : {}", resp);
-    LOGGER.info("ResponseTimes for requestId:{}, table:{}, total time:{}, scanned docs:{}, total docs:{}, scatterGatherStats: {} query: {}", requestId,
-        brokerRequest.getQuerySource().getTableName(), queryProcessingTimeInMillis, resp.getNumDocsScanned(), resp.getTotalDocs(), scatterGatherStats, pql);
-
-    return resp;
+      LOGGER.warn("Validation error on requestId {}: {}", requestId, pql, e);
+      _brokerMetrics.addMeteredTableValue(tableName, BrokerMeter.QUERY_VALIDATION_EXCEPTIONS, 1);
+      return BrokerResponseFactory.getBrokerResponseWithException(DEFAULT_BROKER_RESPONSE_TYPE,
+          QueryException.getException(QueryException.QUERY_VALIDATION_ERROR, e));
+    }
+    if (isTraceEnabled) {
+      brokerRequest.setEnableTrace(true);
+    }
+    if (debugOptions != null) {
+      brokerRequest.setDebugOptions(debugOptions);
+    }
+    brokerRequest.setResponseFormat(ResponseType.BROKER_RESPONSE_TYPE_NATIVE.name());
+    _brokerMetrics.addPhaseTiming(tableName, BrokerQueryPhase.REQUEST_COMPILATION,
```

Comment:
The only difference is that addPhaseTiming takes the time counted outside while timeQueryPhase takes a Callable and count the time inside (IMO not as handy as addPhaseTiming).